### DISCUSSION TITLE
question/two-way-binding

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#15] [Announcements]
# question/two-way-binding

北美前端面试攻略

https://us-fe.github.io/question/two-way-binding.html

**Comment by xiaoyu2er** [comment]
```js
function model(state: {value: string}, element: HTMLInputElement) {
  element.value = state.value;

  element.addEventListener('change', (e) => {
    state.value = e.target.value;
  });

  Object.defineProperty(state, 'value', {
    get() {
      return element.value;
    },
    set(val) {
      element.value = val;
    }
  })
}


```

**Reply by xiaoyu2er** [reply]
关联知识点 
Creating and triggering events
 https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events

Object.defineProperty()
https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

**Comment by mengligit** [comment]
我的get() 开始写的时候是  return state.value;
然后运行到  console.log(state.value) // 'BFE.dev'

的时候就会不停的call get() 。。。。。但是我不太懂这是为啥。。。我试着把get()里面改成this.value, 还是一样
有大神能给我讲讲为啥吗

**Reply by xiaoyu2er** [reply]
第一个问题：return state.value; 为什么会在 `console.log(state.value) // 'BFE.dev'` 的时候就会不停的 `call get() `

回答：

因为你定义的就是 `Object.defineProperty(state, 'value', { get } ) ` 呀，相当于循环调用了

**Reply by xiaoyu2er** [reply]
第二个问题： 我试着把 `get()`里面改成`this.value`, 还是一样

回答：
```js
function model(state , element) {
  element.value = state.value;

  element.addEventListener('change', (e) => {
    state.value = e.target.value;
  });

  Object.defineProperty(state, 'value', {
    get() {
        console.log(this === state)
      return element.value;
    },
    set(val) {
      element.value = val;
    }
  })
}
const div = document.createElement('div')
const state = {value: 1}
state.value
// true 
// 1
```

说明在 defineProperty 这个函数中的 get set 中 this 指向 第一个参数。

**Reply by mengligit** [reply]
感谢，我明白了。假如跳出这道题，没有双向绑定的时候，需要写get返回一个property的值，感觉还要另外定义一个property, 以避免出现不停的调用的这个问题
